Deserialization derive macros

wprzytula opened the pull request:
Fixes #962

Derive macros for `DeserializeValue` and `DeserializeRow` are introduced. 

## An example feature-complete use

```rust
#[derive(scylla_macros::DeserializeValue, PartialEq, Eq, Debug)]
#[scylla(enforce_order, skip_name_checks, forbid_excess_udt_fields)]
struct Udt<'a> {
    a: &'a str,
    #[scylla(skip)]
    x: String,
    #[scylla(allow_missing)]
    b: Option<i32>,
    #[scylla(rename = "d", default_when_null)]
    c: i64,
}

#[derive(scylla_macros::DeserializeRow, PartialEq, Eq, Debug)]
#[scylla(enforce_order)
struct MyRow<'a> {
    a: &'a str,
    #[scylla(skip)]
    x: String,
    b: Option<i32>,
    #[scylla(rename = "d")]
    c: i64,
}
```

## Features

### Flavours
Both macros support both _unordered_ (the default) and _enforce_order_ flavours. These are analogous to `Serialize{Value,Row}` macros flavours, i.e.
- _unordered_ matches by name and allows for arbitrary order of fields in Rust struct and of fields in CQL UDT or columns in CQL row;
- _enforce_order_ matches by position, verifying that names match (unless `skip_name_checks` attribute is given). This is more strict, but also more performant (because the code, by expecting a specific order, can be more compact and less dynamic).

### Struct attributes specific to `DeserializeValue`:
- `forbid_excess_udt_fields` - by default, if a UDT definition contains more fields than the Rust struct (in _unordered_ flavour: anywhere, in _enforce_order_: in suffix), those excess fields are ignored. With `forbid_excess_udt_fields` attribute set, type check fails with an error in case such fields are present. For comparison, `DeserializeRow` always requires the same number of Rust fields and CQL columns, effectively rejecting rows with excess columns.

### Field attributes common to `DeserializeValue` and `DeserializeRow`
- `skip` - a Rust struct field is skipped during type check and deserialization, and initialised with `Default::default()`;
- `rename = "X"` - a Rust struct field is matched to CQL UDT field / row column with the name specified instead of its own name.

### Field attributes specific to `DeserializeValue`
The following attributes are very important in production when adding new fields to a UDT, if one first updates clients, then the cluster. They allow for seamless transition in such case.

- `allow_missing` - if a Rust struct field's name is missing from CQL UDT definition, the field is initialised with `Default::default()`;
- `default_when_null` - if a CQL UDT field contains null, but the Rust field's type expects non-null, deserialization would normally fail. With this attribute, the field is initialised with `Default::default()` instead.

## Errors
The returned errors are rich and insightful, using the new deserialization error framework. They are well-tested.

## Testing
All flavours and attributes should be tested in various combinations. Please verify that I didn't forget about any.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [ ] I have adjusted the documentation in `./docs/source/`.
- [x] I added appropriate `Fixes:` annotations to PR description.

wprzytula commented:
Cassandra failure in CI is unrelated to the PR. @Lorak-mmk @muzarski will you take a look?

wprzytula commented:
v1.1: more attribute verification & tests for that.

muzarski commented on scylla-cql/src/types/deserialize/value.rs:
````diff
@@ -3704,3 +3704,16 @@ fn _test_udt_bad_attributes_rename_collision_with_field() {}
 /// ```
 fn _test_udt_bad_attributes_rename_collision_with_another_rename() {}
+
+/// ```compile_fail
+///
+/// #[derive(scylla_macros::DeserializeValue)]
+/// #[scylla(crate = crate)]
+/// struct TestUdt {
+///     a: i32,
+///     #[scylla(allow_missing)]
+///     b: bool,
+///     c: String,
+/// }
+/// ```
+fn _test_udt_bad_attributes_name_enforce_order_limitations_on_allow_missing() {}
````
A follow-up question: are such test cases executed somewhere? I'm not familiar with the doc tests. I'm curious since CI passed, even though this test is not correct IMO.

wprzytula replied:
Apparently, all those doc tests fail because of an unresolved import:
```rust
rror[E0433]: failed to resolve: unresolved import
 --> scylla-cql/src/types/deserialize/value.rs:1957:10
  |
7 | #[derive(DeserializeValue)]
  |          ^^^^^^^^^^^^^^^^ unresolved import
  |
```
trying to understand why.

wprzytula replied:
The reason was putting `crate = crate` instead of `crate = scylla_cql`. Doc tests are run from outside the crate's context, that's why it didn't work.

wprzytula replied:
> Why should this fail? I don't see enforce_order and skip_name_checks attributes defined (and this should fail only in this case, as I understood from the commit message).

Good catch, I've added those attributes, along with a "control sample" test which is expected to compile successfully.

muzarski commented on scylla-macros/src/deserialize/row.rs:
```diff
@@ -76,7 +82,10 @@ impl Field {
+        };
         syn::LitStr::new(&field_name, Span::call_site())
     }
 }
```
```rust
        let field_name = match self.rename.as_ref() {
            Some(rename) => rename,
            None => &self.ident.as_ref().unwrap().unraw().to_string(),
        };
        syn::LitStr::new(field_name, Span::call_site())
```

wprzytula replied:
Once attribute validation commit is introduced, your suggestion no longer is applicable. Anyway, this allocation happens only during compile time, so let's leave it as-is.

wprzytula commented:
Rebased on main.

wprzytula commented on scylla-cql/src/types/deserialize/row.rs:
```diff
@@ -707,6 +707,37 @@ mod tests {
+            #[allow(unused)]
+            #[scylla(rename = "b")]
+            a: i32,
+            #[allow(unused)]
```
They are not, good point.

wprzytula commented on scylla-macros/src/deserialize/value.rs:
```diff
@@ -108,11 +115,248 @@ impl StructDesc {
+                    }
+                );
+
+                // Verify that the field count is correct
+                if fields.len() < #required_field_count_lit {
+                    return ::std::result::Result::Err(too_few_fields());
+                }
```
You are probably right, but it costs us nothing to do such an early count check. I wouldn't delete it.

muzarski replied:
Yeah, this is rather a nitpick. I agree, let's not delete it.

wprzytula commented on scylla-macros/src/deserialize/value.rs:
```diff
@@ -108,11 +115,248 @@ impl StructDesc {
+            } else {
+                #name_mismatch
+            }
+        };
```
I don't see a cause of that discrepancy. Would you prefer to make them consistent? Which one should I change then?

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -2542,6 +2547,81 @@ pub(super) mod tests {
     }
 
+    #[test]
+    fn test_udt_strict_ordering() {
```
Great that you mentioned this, because a test has detected a bug that would make deserialization panic unconditionally upon excess fields.

wprzytula replied:
Added the tests you asked for.

wprzytula commented:
v1.2: addressed most of the review comments by @muzarski. Thank you very much for spotting bugs!

wprzytula commented:
> > v1.2: addressed most of the review comments by @muzarski. Thank you very much for spotting bugs!
> 
> Remember that previously, the doc tests were not executed in CI (the CI passed, even though there was a buggy test). Please, make sure that we include doc tests in CI. I briefly checked, and see that there is no `cargo test --doc` step in CI.

The doc tests have always been executed in CI. The CI passed even though the test was buggy, because the tests failed to compile (as expected by the tests) because of different reasons than expected.

`cargo test` runs a superset of tests run by `cargo test --doc`.

muzarski commented:
> > > v1.2: addressed most of the review comments by @muzarski. Thank you very much for spotting bugs!
> > 
> > 
> > Remember that previously, the doc tests were not executed in CI (the CI passed, even though there was a buggy test). Please, make sure that we include doc tests in CI. I briefly checked, and see that there is no `cargo test --doc` step in CI.
> 
> The doc tests have always been executed in CI. The CI passed even though the test was buggy, because the tests failed to compile (as expected by the tests) because of different reasons than expected.
> 
> `cargo test` runs a superset of tests run by `cargo test --doc`.

Oh, makes sense. I forgot that all tests failed because of `crate` attribute value.

Lorak-mmk commented on scylla-cql/src/types/deserialize/row.rs:
````diff
@@ -1325,3 +1354,47 @@ mod tests {
+/// struct TestRow {}
+/// ```
+fn _test_struct_deserialization_name_check_skip_requires_enforce_order() {}
+
````
Is the only purpose of those empty test functions describing what the test does? If so, would a comment be better?

wprzytula replied:
Hmm, the purpose of those functions is that they make the associated doctests run by the test runner. And their names are descriptive so that if a test fails, we instantly see, looking at its name, what feature has a fault.

Lorak-mmk replied:
Ok, makes sense.

Lorak-mmk commented on scylla-cql/src/types/deserialize/value.rs:
Regarding commit 

DeserializeValue: allow_missing usage validation 

I don't think nondeterministic is the right word here. Unsolvable? Intractable?

wprzytula replied:
I'll change the word to _unsolvable_.

Lorak-mmk reviewed (changes requested):
Very nice, but big PR, took some time to review.

wprzytula commented on scylla-macros/src/deserialize/value.rs:
```diff
@@ -105,11 +112,248 @@ impl StructDesc {
+                let next_cql_field = cql_field_iter.next()
+                    .map(|(specs, value_res)| value_res.map(|value| (specs, value)))
+                    // Type check has ensured that there are enough CQL UDT fields.
+                    .expect("Too few CQL UDT fields - type check should have prevented this scenario!")
```
OTOH, type check is done only once per query result, whereas `deserialize` is called for each row. So if `deserialize` is at least a bit faster because of not having at least part of type checking logic inside of it, the split is a win.

wprzytula commented:
Rebased on main.

wprzytula commented on scylla-macros/src/deserialize/value.rs:
```diff
@@ -25,6 +25,12 @@ struct Field {
     // with Default::default() instead. All other attributes are ignored.
     #[darling(default)]
     skip: bool,
+
+    // If set, then deserializes from the UDT field with this particular name
+    // instead of the Rust field name.
+    #[darling(default)]
+    rename: Option<String>,
+
```
I believe the cases you mention are tested:
```rust
/// ```compile_fail
///
/// #[derive(scylla_macros::DeserializeValue)]
/// #[scylla(crate = scylla_cql)]
/// struct TestUdt {
///     #[scylla(rename = "b")]
///     a: i32,
///     b: String,
/// }
/// ```
fn _test_udt_bad_attributes_rename_collision_with_field() {}

/// ```compile_fail
///
/// #[derive(scylla_macros::DeserializeValue)]
/// #[scylla(crate = scylla_cql)]
/// struct TestUdt {
///     #[scylla(rename = "c")]
///     a: i32,
///     #[scylla(rename = "c")]
///     b: String,
/// }
/// ```
fn _test_udt_bad_attributes_rename_collision_with_another_rename() {}
```

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -2413,6 +2413,149 @@ pub(super) mod tests {
+                .field(&42_i32.to_be_bytes())
+                .field(&2137_i64.to_be_bytes())
+                .finalize();
+            let typ = udt_def_with_fields([
```
I'll only rename it in the commit which moves test to `value_test.rs`, not to waste time on editing multiple earlier commits.

wprzytula commented on scylla/src/macros.rs:
````diff
@@ -308,6 +308,75 @@ pub use scylla_cql::macros::SerializeRow;
+///
+/// A table defined like this:
+///
+/// ```text
+/// CREATE TABLE ks.my_table (a PRIMARY KEY, b text, c blob);
+/// ```
+///
+/// ...can be deserialized using the following struct:
+///
+/// ```rust
+/// # use scylla_cql::macros::DeserializeRow;
+/// #[derive(DeserializeRow)]
+/// # #[scylla(crate = "scylla_cql")]
+/// struct MyRow<'a> {
+///     a: i32,
+///     b: Option<String>,
+///     c: &'a [u8],
+/// }
+/// ```
````
Please check if the new formulation is OK for you.

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -2586,6 +2591,97 @@ pub(super) mod tests {
+                    b: None,
+                }
+            );
+        }
```
Missing UDT field at the end is equivalent to a serialized null value. That is, it should be deserializable to an Option. For any Default type to be accepted, a `default_when_null` attribute would have to be added.

wprzytula commented on scylla-cql/src/types/deserialize/row.rs:
```diff
@@ -324,6 +392,12 @@ impl Display for BuiltinTypeCheckErrorKind {
                 f,
                 "mismatched types in column {column_name} at index {column_index}: {err}"
             ),
+            BuiltinTypeCheckErrorKind::DuplicatedColumn { column_name, column_index } => write!(
+                f,
+                "column {} occurs more than once in DB metadata; second occurence is at column index {}",
+                column_name,
+                column_index,
+            ),
```
I don't see a way to support it for unordered flavour.

wprzytula commented:
v1.3: addressed @Lorak-mmk comments.

muzarski reviewed (approved):
Sorry, I totally forgot about my comment regarding panicking in case of a name mismatch. I think that we should do the check in `DeserializeRow::deserialize`, and panic just like in `DeserializeValue::deserialize` for udts. Otherwise, LGTM. Good job!

wprzytula commented:
v1.3.1: added a field-column name check in `DeserializeRow` in `enforce_order` flavour.
